lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api-directives-processor.php

```diff
@@ -0,0 +1,162 @@
+<?php
+/**
+ * Interactivity API: WP_Interactivity_API_Directives_Processor class.
+ *
+ * @package WordPress
+ * @subpackage Interactivity API
+ */
+
+if ( class_exists( 'WP_Interactivity_API_Directives_Processor' ) ) {
+	return;
+}
+
+/**
+ * Class used to iterate over the tags of an HTML string and help process the
+ * directive attributes.
+ */
+class WP_Interactivity_API_Directives_Processor extends Gutenberg_HTML_Tag_Processor_6_5 {
+	/**
+	 * Returns the content between two balanced tags.
+	 *
+	 * @return string|null The content between the current opening and its matching closing tag or null if it doesn't
+	 *                     find the matching closing tag.
+	 */
+	public function get_content_between_balanced_tags() {
+		$bookmarks = $this->get_balanced_tag_bookmarks();
+		if ( ! $bookmarks ) {
+			return null;
+		}
+		list( $start_name, $end_name ) = $bookmarks;
+
+		$start = $this->bookmarks[ $start_name ]->start + $this->bookmarks[ $start_name ]->length + 1;
+		$end   = $this->bookmarks[ $end_name ]->start;
+
+		$this->seek( $start_name );
+		$this->release_bookmark( $start_name );
+		$this->release_bookmark( $end_name );
+
+		return substr( $this->html, $start, $end - $start );
+	}
+
+	/**
+	 * Sets the content between two balanced tags.
+	 *
+	 * @param string $new_content The string to replace the content between the matching tags.
+	 * @return bool Whether the content was successfully replaced.
+	 */
+	public function set_content_between_balanced_tags( string $new_content ): bool {
+		$this->get_updated_html();
+
+		$bookmarks = $this->get_balanced_tag_bookmarks();
+		if ( ! $bookmarks ) {
+			return false;
+		}
+		list( $start_name, $end_name ) = $bookmarks;
+
+		$start = $this->bookmarks[ $start_name ]->start + $this->bookmarks[ $start_name ]->length + 1;
+		$end   = $this->bookmarks[ $end_name ]->start;
+
+		$this->seek( $start_name );
+		$this->release_bookmark( $start_name );
+		$this->release_bookmark( $end_name );
+
+		$this->lexical_updates[] = new Gutenberg_HTML_Text_Replacement_6_5( $start, $end - $start, esc_html( $new_content ) );
+		return true;
+	}
+
+	/**
+	 * Returns a pair of bookmarks for the current opening tag and the matching
+	 * closing tag.
+	 *
+	 * @return array|null A pair of bookmarks, or null if there's no matching closing tag.
+	 */
+	private function get_balanced_tag_bookmarks() {
+		$i = 0;
+		while ( array_key_exists( 'start' . $i, $this->bookmarks ) ) {
+			++$i;
+		}
+		$start_name = 'start' . $i;
+
+		$this->set_bookmark( $start_name );
```

Comment:
```suggestion
		static $i = 0;
		$start_name = 'start' . ++$i;
```

Reply:
note that we may still have what should be an extremely unlikely reason to check if the array key exists, because hypothetically something could set one of these bookmarks from the outside. we can use `isset()` safely because a bookmark won't exist in the array if it's not there.

probably find simply to omit the loop, but even if we keep it in the static var will help us avoid walking through all of the bookmarks all of the time since we won't have to start at zero every time.

Reply:
Done in https://github.com/WordPress/gutenberg/pull/58066/commits/4ca55a5504f0c0080ddf88ee745439bd3a8398c6. I also removed the bookmarks to make them less unlikely to collide.